### PR TITLE
[RFC FS-1069] Implicit yields (allow dropping yield in list, array, sequence and computation expressions)

### PR DESCRIPTION
**Pull request by dsyme**
Continuation of #6304 from feature branch

This is the implemntation for [RFC FS-1069 Implict yields](https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1069-implicit-yields.md), i.e. the ["implicit yields" language suggestion](https://github.com/fsharp/fslang-suggestions/issues/643). This allows implicit yield in list, array, sequence and those computation expressions supporting Yield/Combine/Zero/Delay.

See the RFC for notes and discussion

### CODE REVIEW

File: tests/fsharp/typecheck/sigs/neg24.fs [hunk 1]
```diff
@@ -69,3 +69,245 @@ type C() =
...
+        }
+
+    let v3bc = 
+        list {
+          printfn "hello"
+          if true then yield 1 else yield 2 // equivalent explicit yield
+          yield! empty
+        }
+
+    // Note, no subsumption is permitted when using yield in computation expressions, with or eithout implicit yield
```
Comment (cartermp) [inline tests/fsharp/typecheck/sigs/neg24.fs]: Typo: `eithout` --> `without`

Comment (cartermp) [reply tests/fsharp/typecheck/sigs/neg24.fs]: Not sure how I feel about demonstrating this via a comment here. It almost implies that we're ignoring a test case

File: src/fsharp/FSComp.txt [hunk 2]
```diff
@@ -572,7 +572,7 @@ tcUseWhenPatternGuard,"Character range matches have been removed in F#. Consider
 736,tcExprUndelayed,"TcExprUndelayed: delayed"
 737,tcExpressionRequiresSequence,"This expression form may only be used in sequence and computation expressions"
 738,tcInvalidObjectExpressionSyntaxForm,"Invalid object expression. Objects without overrides or interfaces should use the expression form 'new Type(args)' without braces."
-739,tcInvalidObjectSequenceOrRecordExpression,"Invalid object, sequence or record expression"
+#739,tcInvalidObjectSequenceOrRecordExpression,"Invalid object, sequence or record expression"
```
Comment (cartermp) [inline src/fsharp/FSComp.txt]: What is the significance of the `#`?

File: src/fsharp/TypeChecker.fs [hunk 3]
```diff
@@ -6069,6 +6047,19 @@ and TcExprUndelayed cenv overallTy env tpenv (synExpr: SynExpr) =
             let expr2, tpenv = TcStmtThatCantBeCtorBody cenv env tpenv synExpr2
             Expr.Sequential (expr1, expr2, ThenDoSeq, sp, m), tpenv
 
+    // Used to implement the type-directed 'implicit yield' rule for computation expressions
+    | SynExpr.SequentialOrImplicitYield (sp, synExpr1, synExpr2, otherExpr, m) ->
+        let isStmt, e1', tpenv = TryTcStmt cenv env tpenv synExpr1
```
Comment (cartermp) [inline src/fsharp/TypeChecker.fs]: Nit: rename `e1'` so it can show in the debugger

File: src/fsharp/TypeChecker.fs [hunk 4]
```diff
@@ -6069,6 +6047,19 @@ and TcExprUndelayed cenv overallTy env tpenv (synExpr: SynExpr) =
...
+        if isStmt then 
+            let env = ShrinkContext env m synExpr2.Range
+            let e2', tpenv = TcExprThatCanBeCtorBody cenv overallTy env tpenv synExpr2
```
Comment (cartermp) [inline src/fsharp/TypeChecker.fs]: Nit: rename `e2'` so it can show in the debugger

File: src/fsharp/FSComp.txt (hunk 2 above)
Comment (KevinRansom) [reply src/fsharp/FSComp.txt]: It's commented out.

**Comment by KevinRansom** [comment]
@dsyme, is this ready to go?

File: src/fsharp/FSComp.txt (hunk 2 above)
Comment (cartermp) [reply src/fsharp/FSComp.txt]: This should not be commented out. We should either remove it or keep it.

**Comment by dsyme** [comment]
@KevinRansom Yes this is ready

**Comment by dsyme** [comment]
@KevinRansom I have to address the code review

File: tests/fsharp/typecheck/sigs/neg24.fs (hunk 1 above)
Comment (dsyme) [reply tests/fsharp/typecheck/sigs/neg24.fs]: I added this as a negative test case, thanks

**Comment by dsyme** [comment]
This is ready up to getting it green

**Comment by cartermp** [comment]
@dsyme test failure here: https://dev.azure.com/dnceng/public/_build/results?buildId=216204&view=ms.vss-test-web.build-test-results-tab

(side note: love that @brettfo has this level of integration with our tests set up!)

**Comment by dsyme** [comment]
> side note: love that @brettfo has this level of integration with our tests set up!

@brettfo That is amazing!!

@cartermp Shoud be fixed by latest push

**Comment by cartermp** [comment]
It's not clear why the 7k line FSComp file is being included here. Is this a clean diff?

**Comment by dsyme** [comment]
> It's not clear why the 7k line FSComp file is being included here. Is this a clean diff?


I cleaned up the diff, thanks

**Comment by dsyme** [comment]
> It also introduces names that are not debuggable; these should change.

What were you referring to for this? thanks!

**Comment by dsyme** [comment]
We had a transient test failure here, of a kind I've not seen before
```
Error Message:
 System.ObjectDisposedException : The CancellationTokenSource has been disposed.
Stack Trace:
   at System.Threading.CancellationTokenSource.ThrowObjectDisposedException()
   at FSharp.Compiler.SourceCodeServices.Reactor.SetBackgroundOp(FSharpOption`1 bgOpOpt) in D:\a\1\s\src\fsharp\service\Reactor.fs:line 137
   at FSharp.Compiler.SourceCodeServices.BackgroundCompiler.CheckProjectInBackground(FSharpProjectOptions options, String userOpName) in D:\a\1\s\src\fsharp\service\service.fs:line 827
   at FSharp.Compiler.SourceCodeServices.FSharpChecker.StartBackgroundCompile(FSharpProjectOptions options, FSharpOption`1 userOpName) in D:\a\1\s\src\fsharp\service\service.fs:line 1165
   at Microsoft.VisualStudio.FSharp.LanguageService.FSharpLanguageServiceBackgroundRequests_DEPRECATED.ExecuteBackgroundRequest(FSharpBackgroundRequest_DEPRECATED req, IFSharpSource_DEPRECATED source) in D:\a\1\s\vsintegration\src\FSharp.LanguageService\BackgroundRequests.fs:line 269
   at Salsa.Salsa.Privates.SimpleOpenFile.TryExecuteBackgroundRequest(FSharpBackgroundRequest_DEPRECATED pr) in D:\a\1\s\vsintegration\tests\Salsa\Salsa.fs:line 1129
   at Salsa.Salsa.Privates.SimpleOpenFile.OnIdleTypeCheck() in D:\a\1\s\vsintegration\tests\Salsa\Salsa.fs:line 1158
   at Salsa.Salsa.Privates.SimpleOpenFile.OnIdle() in D:\a\1\s\vsintegration\tests\Salsa\Salsa.fs:line 1115
   at Salsa.Salsa.Privates.SimpleVisualStudio.Salsa-Salsa-IVisualStudio-OnIdle() in D:\a\1\s\vsintegration\tests\Salsa\Salsa.fs:line 848
   at Salsa.Salsa.Privates.SimpleVisualStudio.Salsa-Salsa-IVisualStudio-TakeCoffeeBreak() in D:\a\1\s\vsintegration\tests\Salsa\Salsa.fs:line 856
   at UnitTests.TestLib.LanguageService.LanguageServiceBaseTests.CreateSingleFileProject(FSharpList`1 content, FSharpOption`1 references, FSharpOption`1 defines, FSharpOption`1 fileKind, FSharpOption`1 disabledWarnings, FSharpOption`1 fileName) in D:\a\1\s\vsintegration\tests\UnitTests\TestLib.LanguageService.fs:line 325
   at UnitTests.TestLib.LanguageService.LanguageServiceBaseTests.CreateSingleFileProject(String content, FSharpOption`1 references, FSharpOption`1 defines, FSharpOption`1 fileKind, FSharpOption`1 disabledWarnings, FSharpOption`1 fileName) in D:\a\1\s\vsintegration\tests\UnitTests\TestLib.LanguageService.fs:line 287
   at Tests.LanguageService.AutoCompletion.UsingMSBuild.VerifyDotCompListContainAllAtStartOfMarker(String fileContents, String marker, FSharpList`1 list, FSharpOption`1 addtlRefAssy, FSharpOption`1 coffeeBreak)
   at Tests.LanguageService.AutoCompletion.UsingMSBuild.QueryExpression.DotCompletionSystematic1() in D:\a\1\s\vsintegration\tests\UnitTests\LegacyLanguageService\Tests.LanguageService.Completion.fs:line 7604
```

**Comment by dsyme** [comment]
@KevinRansom @TIHan Please advise on adding the /langversion checks for this PR. This needs to be done asap, thanks. I can do it, just let me know.

**Comment by KevinRansom** [comment]
@dsyme, there appears to be some cancellation token issue with Linux.  A bunch of meregs have failed because of it..

**Comment by KevinRansom** [comment]
@dsyme, I had planned to work through all of the PR's doing the langversion change to fgharp47 and fsharp5 is broken at the moment due to some issue with cancellationtokens and linux.  Have you seen this failure on a platform other than linux?

**Comment by KevinRansom** [comment]
Merging, I will add in the langversion support post merge.

**Pull request merged**